stimulize.man
.\" Manpage for stimulize.
.\" Create a git pull request at https://github.com/sergiolemus/stimulize 
.TH man 1 "28 March 2019" "1.0" "STIMULIZE User Manual"
.SH NAME
.B stimulize 
\- create 
.I irsim 
command scripts to test logic circuits easily and programmatically.

.SH SYNOPSIS
.B stimulize 
-h[elp]
.br
.B stimulize
-a[nalyzer] {a_nodes} -v[ector] {v_nodes} [options]

.SH DESCRIPTION
.I stimulize
creates an irsim cmd format from given inputs. It will display {a_nodes} on the analyzer window
and will step through input combinations of {v_nodes}. The output of the program is written
to stdout can be piped to a file and then imported to either standalone, terminal
.I irsim
or 
.I GNU Electric irsim 
(available in the SFS extension) easily.

.SH OPTIONS
If no 
.I stepsize 
is given, the default is set to 5(ns). If no test sequence is provided, the program will
automatically step through all combinations of {v_nodes} through gray code sequence.
.br
.PP
.TP
.IR \-a " " a_nodes
List which nodes should be printed on the analyzer window. Arguments should be space seperated
.br
Node lists may be given in format NAME[<num1>:<num2>] and will be auto expanded. See Example 3
.TP
.IR \-h
help
.TP
.IR \-s " " stepsize
step size, in nanoseconds. If no argument is given, defaults to 5ns
.TP
.IR \-t " " test_sequence
Test sequency, in binary. You can feed the program a sequence of N-bit binary numbers, where
N is the number of v_nodes there are. Arguments should be space seperated. 
.br
If no sequence is given, the program will step through every N bit combination through gray code
sequence such that only 1 bit changes during each step.
.TP
.IR \-v " " v_nodes
List which nodes will be controlled as a vector input to the circuit. Arguments should be space seperated
.br
Node lists may be given in format NAME[<num1>:<num2>] and will be auto expanded. See Example 3

.SH EXAMPLES
.PP
.br
Stepsize: default - 5ns, View [A,B,C,OUT] in the analyzer, Vectorize [A,B,C] as inputs, and step through all combinations in
gray code order.
.br
\fI% stimulize \-a A B C OUT \-v A B C\fR
.br
| STIMULIZE v3.01
.br
| FILE GENERATED BY COMMAND:
.br
| stimulize -a A B C OUT -v A B C 
.br
stepsize 5
.br
analyzer A B C OUT 
.br
| VECTOR: [A,B,C] = [0,0,0]
.br
l A B C
.br
s
.br
| VECTOR: [A,B,C] = [0,0,1]
.br
h C 
.br
s
.br
| VECTOR: [A,B,C] = [0,1,1]
.br
h B 
.br
s
.br
| VECTOR: [A,B,C] = [0,1,0]
.br
l C 
.br
s
.br
| VECTOR: [A,B,C] = [1,1,0]
.br
h A 
.br
s
.br
| VECTOR: [A,B,C] = [1,1,1]
.br
h C 
.br
s
.br
| VECTOR: [A,B,C] = [1,0,1]
.br
l B 
.br
s
.br
| VECTOR: [A,B,C] = [1,0,0]
.br
l C 
.br
s
.br
| VECTOR: [A,B,C] = [0,0,0]
.br
l A B C 
.br
s
.PP
Stepsize: 50ns, View [I0,I1,I2,I3,O0,O1] in the analyzer, Vectorize [I0,I1,I2,I3] as inputs, and step through
the specific specific cases:
.br
[I0,I1,I2,I3] = [0,0,0,0], [1,1,1,0], [0,0,1,1], [1,1,1,1].
.br
\fI% stimulize \-a I0 I1 I2 I3 O0 O1 \-v I0 I1 I2 I3 -s 50 -t 0000 1110 0011 1111\fR
.br
| STIMULIZE v3.01
.br
| FILE GENERATED BY COMMAND:
.br
| stimulize -a I0 I1 I2 I3 O0 O1 -v I0 I1 I2 I3 -s 50 -t 0000 1110 0011 1111 
.br
stepsize 50
.br
analyzer I0 I1 I2 I3 O0 O1 
.br
| VECTOR: [I0,I1,I2,I3] = [0,0,0,0]
.br
l I0 I1 I2 I3 
.br
s
.br
| VECTOR: [I0,I1,I2,I3] = [1,1,1,0]
.br
h I0 I1 I2 
.br
s
.br
| VECTOR: [I0,I1,I2,I3] = [0,0,1,1]
.br
l I0 I1 
.br
h I3 
.br
s
.br
| VECTOR: [I0,I1,I2,I3] = [1,1,1,1]
.br
h I0 I1 
.br
s
.PP
Stepsize: 50ns, View [A3,A2,A1,A0,B3,B2,B1,B0,F3,F2,F1,F0] in the analyzer, Vectorize [A3,A2,A1,A0,B3,B2,B1,B0] as inputs, 
and step through specific cases
.br
[A3,A2,A1,A0,B3,B2,B1,B0]=[0,0,0,0,0,0,0,0], [1,1,1,1,0,0,0,0,], [0,0,0,0,1,1,1,1]
.br
\fI%stimulize -a A[3:0] B[3:0] F[3:0] -v A[3:0] B[3:0] -s 50 -t 00000000 11110000 00001111\fR
.br
| STIMULIZE v3.41
.br
| FILE GENERATED BY COMMAND:
.br
| ./stimulize -a A3 A2 A1 A0 B3 B2 B1 B0 F3 F2 F1 F0 -v A3 A2 A1 A0 B3 B2 B1 B0 -s 50 -t 00000000 11110000 00001111 
.br
stepsize 50
.br
analyzer A3 A2 A1 A0 B3 B2 B1 B0 F3 F2 F1 F0 
.br
| VECTOR: [A3,A2,A1,A0,B3,B2,B1,B0] = [0,0,0,0,0,0,0,0]
.br
l A3 A2 A1 A0 B3 B2 B1 B0 
.br
s
.br
| VECTOR: [A3,A2,A1,A0,B3,B2,B1,B0] = [1,1,1,1,0,0,0,0]
.br
h A3 A2 A1 A0 
.br
s
.br
| VECTOR: [A3,A2,A1,A0,B3,B2,B1,B0] = [0,0,0,0,1,1,1,1]
.br
l A3 A2 A1 A0 
.br
h B3 B2 B1 B0 
.br
s

.SH Todo's / Known Bugs
1) Allow test sequences to be written in hex
.br
2) Because default test sequences are stored as ints, systems with 64 or more inputs are not processed correctly
The step to correct this would be to make a function that can take in a raw bool vector and automatically increment
in binary or gray code but it currently seems unnecessary.

.SH AUTHOR
2019 Segio Lemus, Joseph Shaker.
.br
Please make all inquiries or bug reports on the github page.
.br
https://github.com/sergiolemus/stimulize
.PP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
.br
* Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
.br
* Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
.br
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

stimulize.cpp
#include <cinttypes>
#include <cstdlib>
#include <iostream>
#include <string>
#include <regex>
#include <vector>

#define VERSION "v3.41"

// PROGRAM TO VECTORIZE IRSIM INPUTS

// REQUIRED FLAGS:
// --analyzer,-a <nodes...>       :   ENTER SPACE SEPARATED NODE NAMES
// --vector,-v <nodes...>         :   ENTER SPACE SEPARATED NODE NAMES TO VECTORIZE

// OPTIONAL FLAGS:
// --stepsize,-s <val>            :   SET STEPSIZE TO VAL ns, DEFAULT IS 5ns
// --test,-t <val_sequence...>    :   ENTER A TEST SEQUENCE OF VALUES

// EXAMPLE: 2 INPUT AND WITH (A,B) AS INPUTS AND (OUT) AS AN OUTPUT
// (A,B) WILL GO RESPECTIVELY FROM (0,0) -> (0,1) -> (1,0) -> (1,1)
// WITH A STEPSIZE OF 5ns.

// ./stimulize -analyzer A B OUT -v A B

// EXAMPLE: (IN_1,IN_2,IN_3) AS INPUTS AND (OUT_1,OUT_2) AS OUTPUTS
// (IN_1,IN_2,IN_3) WILL GO RESPECTIVELY FROM (0,0,0) -> (0,0,1) -> ... -> (1,1,0) -> (1,1,1)
// WITH A STEPSIZE OF 10ns

//./stimulize -a IN_1 IN_2 IN_3 OUT_A OUT_B -v IN_1 IN_2 IN_3 -s 10

//function to convert a binary number to grays code
//counting sequentially in gray code only toggles 1 bit at a time so should give more stable output shapes
template <class T>
T binaryToGray( T bnum )
{
    return ( ( bnum ) ^ ( bnum >> 1 ) );
}

// FUNCTION TO INSPECT THE STATE VECTOR AS A COMMENT
void printState( const std::vector<std::string> &inputs, int v_nodes, unsigned int number,
                 const std::string &sequence )
{
    std::string binary = "";
    
    std::cout << "| VECTOR: [";
    
    v_nodes--;
    for( int i = 0; i < v_nodes; i++ )
    {
        std::cout << inputs[ i ] << ",";
    }
    
    std::cout << inputs[ v_nodes ] << "] = [";
    
    if( sequence == "" )
    {
        if( number & 1 )
        {
            binary = "1";
        }
        else
        {
            binary = "0";
        }
    
        number = number >> 1;
    
        v_nodes++;
        for( int i = 1; i < v_nodes; i++ )
        {
            if( number & 1 )
            {
                binary = "1," + binary;
            }
            else
            {
                binary = "0," + binary;
            }
            
            number = number >> 1;
        }
        
        std::cout << binary;
    }
    else
    {
        for( int i = 0; i < v_nodes; i++ )
        {
            std::cout << sequence[ i ] << ",";
        }
        
        std::cout << sequence[ v_nodes ];
    }
    
    std::cout << "]\n";
}

// FUNCTION TO GENERATE VECTORIZED INPUTS
//"DRIVER" that is run by main after error checking
int generateCommands( const std::vector<std::string> &argv, int a_index, int a_nodes, 
                       int v_index, int v_nodes, int t_index, int t_nodes, bool t, int s_index )
{

//  auto make_time = std::chrono::system_clock::now();
    std::cout << "| STIMULIZE " << VERSION << "\n"
              << "| FILE GENERATED BY COMMAND:\n"
              << "| ";

    for ( int i = 0; i < argv.size(); i++ )
    {
        std::cout << argv[ i ] << " ";
    }
    std::cout << "\n\n";

    std::vector<std::string> inputs      ( v_nodes );
    std::vector<char>        input_values( v_nodes );
    std::vector<bool>        toggle      ( v_nodes );
    
    bool low = false;
    bool high = false;
    
    if( s_index == -1 )
    {
        std::cout << "stepsize 5\n";
    }
    else
    {
        std::cout << "stepsize " << argv[ s_index + 1 ] << "\n";
    }
    
    a_index++;
    std::cout << "analyzer ";
    for( int i = 0; i < a_nodes; i++ )
    {
        std::cout << argv[ a_index + i ] << " ";
    }
    std::cout << "\n\n" << std::flush;
    
    v_index++;
    for( int i = 0; i < v_nodes; i++ )
    {
        inputs[ i ] = argv[ v_index + i ];
        input_values[ i ] = 0;
        toggle[ i ] = true;
    }
    
    //default sequence (count through all combinations in gray)
    //works by counting up all states as a binary/gray number, then masking and putting in inputs[n]
    if( !t )
    {        
        std::uint_least64_t count_to = 1;

        for ( int i = 0; i < v_nodes; i++ )
        {
            count_to *= 2;
        }

        if ( count_to == 0 )
        {
            std::cerr << "Too many vectorized inputs" << std::endl;
            return 2;
        }

        //print first state
        printState( inputs, v_nodes, 0, "" );
        
        std::cout << "l ";
        for( int n = 0; n < v_nodes; n++ )
        {
            std::cout << inputs[ n ] << " ";
        }
        std::cout << "\ns" << ( count_to == 1 ? "": "\n\n");

        //print sequential states states
        for( auto bnum = 1; bnum < count_to; bnum++ )
        {
            auto gnum = binaryToGray( bnum );

            low = false;
            high = false;

            printState( inputs, v_nodes, gnum, "" );
            
            for( int n = v_nodes - 1; n >= 0; n-- )
            {
                //count in binary
                //if( ( ( bnmum & mask ) >> ( shift_right ) != input_values[ n ] ) )

                //count in gray code
                if( ( gnum & 1 ) != input_values[ n ] )
                {
                    toggle[ n ] = true;
                    
                    if( !low && input_values[ n ] == 1 )
                    {
                        low = true;
                    }
                    else if( !high && input_values[ n ] == 0 )
                    {
                        high = true;
                    }
                }
                else
                {
                    toggle[ n ] = false;
                }

                gnum = gnum >> 1;
            }
            
            if( low )
            {
                std::cout << "l ";
                
                for( int n = 0; n < v_nodes; n++ )
                {
                    if( toggle[ n ] && ( input_values[ n ] == 1 ) )
                    {
                        toggle[ n ] = false;
                        input_values[ n ] = 0;
                        std::cout << inputs[ n ] << " ";
                    }
                }
                
                std::cout << "\n";
            }
            
            if( high )
            {
                std::cout << "h ";
                
                for( int n = 0; n < v_nodes; n++ )
                {
                    if( toggle[ n ] && ( input_values[ n ] == 0 ) )
                    {
                        toggle[ n ] = false;
                        input_values[ n ] = 1;
                        std::cout << inputs[ n ] << " ";
                    }
                }
                
                std::cout << "\n";
            }
            
            std::cout << "s\n\n";
        }
        
        printState( inputs, v_nodes, 0, "" );
        
        //end by making all inputs low
        std::cout << "l ";

        for( int n = 0; n < v_nodes; n++ )
        {
            std::cout << inputs[ n ] << " ";
            input_values[ n ] = 0;
        }
        std::cout << "\ns";
    }
    
    //custom test sequence
    //works by explictly maniplating the input_values[n]
    else
    {
        std::string arg;
        
        t_index++;
        arg = std::string( argv[ t_index ] );
        for( int n = 0; n < v_nodes; n++ )
        {
            //convert from ascii representation of "0"/"1" to "0"/"1"
            input_values[ n ] = arg[ n ] - 0x30;
            
            if( !low && input_values[ n ] == 0 )
            {
                low = true;
            }
            else if( !high && input_values[ n ] == 1 )
            {
                high = true;
            }
        }
        
        printState( inputs, v_nodes, 0, arg );
        
        if( low )
        {
            std::cout << "l ";
            
            for( int n = 0; n < v_nodes; n++ )
            {
                if( input_values[ n ] == 0 )
                {
                    std::cout << inputs[ n ] << " ";
                }
            }
            
            std::cout << "\n";
        }
        
        if( high )
        {
            std::cout << "h ";
            
            for( int n = 0; n < v_nodes; n++ )
            {
                if( input_values[ n ] == 1 )
                {
                    std::cout << inputs[ n ] << " ";
                }
            }
            
            std::cout << "\n";
        }
        
        std::cout << "s" << ( t_nodes == 1 ? "": "\n\n");

        for( int i = 1; i < t_nodes; i++ )
        {
            low = false;
            high = false;
            
            arg = std::string( argv[ t_index + i ] );
            for( int n = 0; n < v_nodes; n++ )
            {
                if( (arg[ n ] - 48) != input_values[ n ] )
                {
                    toggle[ n ] = true;
                    
                    if( !low && input_values[ n ] == 1 )
                    {
                        low = true;
                    }
                    else if( !high && input_values[ n ] == 0 )
                    {
                        high = true;
                    }
                }
                else
                {
                    toggle[ n ] = false;
                }
            }
            
            printState( inputs, v_nodes, 0, arg );
            
            if( low )
            {
                std::cout << "l ";
                
                for( int n = 0; n < v_nodes; n++ )
                {
                    if( toggle[ n ] && input_values[ n ] == 1 )
                    {
                        toggle[ n ] = false;
                        input_values[ n ] = 0;
                        std::cout << inputs[ n ] << " ";
                    }
                }
                
                std::cout << "\n";
            }
            
            if( high )
            {
                std::cout << "h ";
                
                for( int n = 0; n < v_nodes; n++ )
                {
                    if( toggle[ n ] && input_values[ n ] == 0 )
                    {
                        toggle[ n ] = false;
                        input_values[ n ] = 1;
                        std::cout << inputs[ n ] << " ";
                    }
                }
                
                std::cout << "\n";
            }
            
            std::cout << "s" << ( i == ( t_nodes - 1 ) ? "" : "\n\n" );
        }
    }
    std::cout << std::flush;
    return 0;
}

void show_usage()
{
    std::cerr << std::endl
              << "Usage:\n\n"
              << "stimulize -a {node_name ..} -v {node_name ..} "
              << "[-s <step_val>] [-t {test_seq ..}]\n\n" 
    
              << "Required:\n\n" 
              << "  -a,--analyzer {node_name ..}          "
              << "nodes to view on the analyzer window.\n"
              << "  -v,--vector {node_name ..}            "
              << "nodes to vectorize.\n" 
    
              << "Optional:\n" 
              << "  -s,--stepsize <step_val>              "
              << "set the stepsize to <step_val>, default is 5ns.\n" 
              << "  -t,--test {test_seq ..}               "
              << "binary sequences to test on vector.\n"
              << "                                        "
              << "default is all combinations through gray code\n\n" 

              << "Examples:\n" 
              << "  stepsize: 5ns, analyzer: [A,B,C,OUT], vector: [A,B,C], all combinations, gray code\n\n"
              << "$ stimulize -a A B C OUT -v A B\n"
              << "  stepsize: 50ns, analyzer: [I0,I1,I2,I3,O0,O1], vector: [I0,I1,I2,I3] = [0,0,0,0] [1,1,1,0] [0,0,1,1] [1,1,1,1]\n"
              << "$ stimulize -a I0 I1 I2 I3 O0 O1 -v I0 I1 I2 I3 -t 0000  1110  0011 1111 > test.cmd\n\n"
              << "If this program isinstalled on your system, use \n"
              << "$ man stimulize\n"
              << "for more / better formatted details\n\n" << std::flush;
}

//error checks inputs, then passes to "generateCommands"
int main( int argc, char** argv )
{
    int analyzer_index = -1;
    int analyzer_nodes = 0;
    bool analyzer = false;
    
    int vector_index = -1;
    int vector_nodes = 0;
    bool vector = false;
   
    int stepsize_index = -1;
    int stepsize_nodes = 0;
    bool stepsize = false;
    
    int test_index = -1;
    int test_nodes = 0;
    bool test = false;
    
    std::string flag = "";
    std::string arg;
    int count;
    
    bool error = false;
    
    //arguments are transfered from argv to argv_v to allow for expansion
    //count and store how many arguments there are for analyzer nodes, v nodes, etc
    std::vector<std::string> argv_v;
    
    std::regex expand(".*\\[\\d+:\\d+\\]");
    std::regex number("\\d+");

    if( argc == 1 )
    {
        show_usage( );
        exit( 1 );
    }
    argv_v.push_back( argv[ 0 ] );

    //walk through each argument, verify if it is a flag or another argument,
    //update indicies for analyzer,vector,step,test, etc, and copy all to argv_v
    for( int i = 1; i < argc; i++ )
    {

        arg = std::string( argv[ i ] );

        //if input is of form NAME[0:100]
        //expand to  NAME0 NAME1 .... NAME1000
        if ( std::regex_match( arg, expand ) )
        {
            
            //start looking after the first [ for numbers
            int bracket_index = arg.find( "[" );
            std::string search_from = arg.substr( bracket_index );
            std::sregex_iterator next(search_from.begin(), search_from.end(), number);
            
            std::smatch match = *next;
            int first = std::stoi( match.str() );

            next++;
            match = *next;
            
            int second = std::stoi( match.str() );

            std::string name = arg.substr( 0, bracket_index );
            if ( first <= second )
            {
                for ( int i = first; i <= second; i++ )
                {
                    argv_v.push_back( name + std::to_string( i ) );
                    count++;
                }
            }

            else
            {
                for ( int i = first; i >= second; i-- )
                {
                    argv_v.push_back( name + std::to_string( i ) );
                    count++;
                }
            }
            count--;
        }
        else
        {
            argv_v.push_back( arg );
        }

        if( arg == "-a" || arg == "--analyzer" )
        {
            analyzer_index = argv_v.size() - 1;
            
            if( analyzer == true )
            {
                std::cerr << "--analyzer, -a: Ocurred too many times\n";
                error = true;
            }
            
            analyzer = true;
            
            if( flag == "v" )
            {
                vector_nodes = count;
            }
            else if( flag == "t" )
            {
                test_nodes = count;
            }
            else if( flag == "s" )
            {
                stepsize_nodes = count;
            }
            
            flag = "a";
            count = -1;
        }
        else if( arg == "-v" || arg == "--vector" )
        {
            vector_index = argv_v.size() - 1;
            
            if( vector == true )
            {
                std::cerr << "--vector, -v: Ocurred too many times\n";
                error = true;
            }
            
            vector = true;
            
            if( flag == "a" )
            {
                analyzer_nodes = count;
            }
            else if( flag == "t" )
            {
                test_nodes = count;
            }
            else if( flag == "s" )
            {
                stepsize_nodes = count;
            }
            
            flag = "v";
            count = -1;
        }
        else if( arg == "-t" || arg == "--test" )
        {
            test_index = argv_v.size() - 1;
            
            if( test == true )
            {
                std::cerr << "--test, -t: Ocurred too many times\n";
                error = true;
            }
            
            test = true;
            
            if( flag == "a" )
            {
                analyzer_nodes = count;
            }
            else if( flag == "v" )
            {
                vector_nodes = count;
            }
            else if( flag == "s" )
            {
                stepsize_nodes = count;
            }
            
            flag = "t";
            count = -1;
        }
        else if( arg == "-s" || arg == "--stepsize" )
        {
            stepsize_index = argv_v.size() - 1;
            
            if( stepsize == true )
            {
                std::cerr << "--stepsize, -s: Ocurred too many times\n";
                error = true;
            }
            
            stepsize = true;
            
            if( flag == "a" )
            {
                analyzer_nodes = count;
            }
            else if( flag == "v" )
            {
                vector_nodes = count;
            }
            else if( flag == "t" )
            {
                test_nodes = count;
            }
            
            flag = "s";
            count = -1;
        }
        else if( arg[ 0 ] == '-' )
        {
            show_usage();
            return 1;
        }
        count++;
    }
    
    if( flag == "v" )
    {
        vector_nodes = count;
    }
    else if( flag == "t" )
    {
        test_nodes = count;
    }
    else if( flag == "a" )
    {
        analyzer_nodes = count;
    }
    else if( flag == "s" )
    {
        stepsize_nodes = count;
    }
    
    if( !analyzer )
    {
        std::cerr << "--analyzer, -a: Flag required.\n";
        error = true;
    }
    
    if( !vector )
    {
        std::cerr << "--vector, -v: Flag required.\n";
        error = true;
    }
    
    if( analyzer && analyzer_nodes == 0 )
    {
        std::cerr << "--analyzer, -a: No nodes submitted.\n";
        error = true;
    }
    
    if( vector && vector_nodes == 0 )
    {
        std::cerr << "--vector, -v: No nodes submitted.\n";
        error = true;
    }
    
    bool non_binary  = false;
    
    //if running a custom test sequence, verify it is valid
    if( test )
    {
        if( test_nodes == 0 )
        {
            std::cerr << "--test, -t: No test sequence submitted.\n";
            error = true;
        }
    
        for( int i = 0; i < test_nodes; i++ )
        {
            arg = std::string( argv_v[ (test_index + 1) + i ] );
            if( arg.length() != vector_nodes )
            {
                std::cerr << "--test, -t: Sequence (" << i+1 << ")";
                std::cerr << " contains (" << arg.length() << "), not (" << vector_nodes << ") values: ";
                std::cerr << arg << '\n';
                error = true;
            }
            
            non_binary = false;
            
            for( int c = 0; c < arg.length(); c++ )
            {
                if( arg[ c ] != '0' && arg[ c ] != '1' )
                {
                    if( !non_binary )
                    {
                        std::cerr << "--test, -t: Sequence (" << i+1 << ")";
                        std::cerr << " contains non-binary character(s): ";
                        
                        non_binary = true;
                    }
                    
                    std::cerr << arg[c];
                    error = true;
                }
            }
            
            if( non_binary )
            {
                std::cerr << std::endl;
            }
        }
    }
    
    if(  stepsize && stepsize_nodes == 0 )
    {
        std::cerr << "--stepsize, -s: No value submitted.\n";
        error = true;
    }
    
    if( stepsize && stepsize_nodes > 1 )
    {
        std::cerr << "--stepsize, -s: More than one value was passed in.\n";
        error = true;
    }
    
    if( error )
    {
        std::cerr << "If this program is installed on your system, use \n"
                  << "$ man stimulize\n"
                  << "for more details\n" << std::flush;
        return 1;
    }
    
    return generateCommands( argv_v, analyzer_index, analyzer_nodes, vector_index, vector_nodes,
                     test_index, test_nodes, test, stepsize_index );
    
}